// vscode-azureextensionui/out/src/extensionVariables.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require("assert");
const localize_1 = require("./localize");
class UninitializedExtensionVariables {
    constructor() {
        this._error = new Error(localize_1.localize('uninitializedError', '"registerUIExtensionVariables" must be called before using the vscode-azureextensionui package.'));
    }
    get context() {
        throw this._error;
    }
    get outputChannel() {
        throw this._error;
    }
    get ui() {
        throw this._error;
    }
    get reporter() {
        throw this._error;
    }
}
/**
 * Container for common extension variables used throughout the UI package. They must be initialized with registerUIExtensionVariables
 */
exports.ext = new UninitializedExtensionVariables();
function registerUIExtensionVariables(extVars) {
    assert(extVars.context, 'registerUIExtensionVariables: Missing context');
    assert(extVars.outputChannel, 'registerUIExtensionVariables: Missing outputChannel');
    assert(extVars.reporter, 'registerUIExtensionVariables: Missing reporter');
    assert(extVars.ui, 'registerUIExtensionVariables: Missing ui');
    exports.ext = extVars;
}
exports.registerUIExtensionVariables = registerUIExtensionVariables;
