// microsoft.aspnetcore.razor.vscode/dist/RazorDocumentChangeKind.js
"use strict";
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
Object.defineProperty(exports, "__esModule", { value: true });
var RazorDocumentChangeKind;
(function (RazorDocumentChangeKind) {
    RazorDocumentChangeKind[RazorDocumentChangeKind["added"] = 0] = "added";
    RazorDocumentChangeKind[RazorDocumentChangeKind["removed"] = 1] = "removed";
    RazorDocumentChangeKind[RazorDocumentChangeKind["opened"] = 2] = "opened";
    RazorDocumentChangeKind[RazorDocumentChangeKind["closed"] = 3] = "closed";
    RazorDocumentChangeKind[RazorDocumentChangeKind["csharpChanged"] = 4] = "csharpChanged";
    RazorDocumentChangeKind[RazorDocumentChangeKind["htmlChanged"] = 5] = "htmlChanged";
})(RazorDocumentChangeKind = exports.RazorDocumentChangeKind || (exports.RazorDocumentChangeKind = {}));
