// vscode-azureextensionui/out/src/treeDataProvider/AzureTreeItem.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable-next-line:no-require-imports
const opn = require("opn");
const errors_1 = require("../errors");
const localize_1 = require("../localize");
const treeConstants_1 = require("./treeConstants");
class AzureTreeItem {
    constructor(parent) {
        this.parent = parent;
    }
    get _effectiveDescription() {
        return this._temporaryDescription || this.description;
    }
    get fullId() {
        let id = this.id || this.label;
        if (!id.startsWith('/')) {
            id = `/${id}`;
        }
        // For the sake of backwards compat, only add the parent's id if it's not already there
        if (this.parent && !id.startsWith(this.parent.fullId)) {
            id = `${this.parent.fullId}${id}`;
        }
        return id;
    }
    get effectiveIconPath() {
        return this._temporaryDescription ? treeConstants_1.loadingIconPath : this.iconPath;
    }
    get effectiveLabel() {
        return this._effectiveDescription ? `${this.label} (${this._effectiveDescription})` : this.label;
    }
    get root() {
        if (this.parent) {
            return this.parent.root;
        }
        else {
            throw new errors_1.ArgumentError(this);
        }
    }
    get treeDataProvider() {
        if (this.parent) {
            return this.parent.treeDataProvider;
        }
        else {
            throw new errors_1.ArgumentError(this);
        }
    }
    //#endregion
    refresh() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.refreshImpl) {
                yield this.refreshImpl();
            }
            if (this.refreshLabelImpl) {
                yield this.refreshLabelImpl();
            }
            yield this.treeDataProvider.refresh(this);
        });
    }
    openInPortal(id, options) {
        id = id === undefined ? this.fullId : id;
        const queryPrefix = (options && options.queryPrefix) ? `?${options.queryPrefix}` : '';
        const url = `${this.root.environment.portalUrl}/${queryPrefix}#@${this.root.tenantId}/resource${id}`;
        // tslint:disable-next-line:no-floating-promises
        opn(url);
    }
    includeInTreePicker(expectedContextValues) {
        return expectedContextValues.some((val) => {
            return this.contextValue === val ||
                !this.isAncestorOfImpl ||
                this.isAncestorOfImpl(val);
        });
    }
    deleteTreeItem() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.runWithTemporaryDescription(localize_1.localize('deleting', 'Deleting...'), () => __awaiter(this, void 0, void 0, function* () {
                if (this.deleteTreeItemImpl) {
                    yield this.deleteTreeItemImpl();
                    if (this.parent) {
                        yield this.parent.removeChildFromCache(this);
                    }
                }
                else {
                    throw new errors_1.NotImplementedError('deleteTreeItemImpl', this);
                }
            }));
        });
    }
    runWithTemporaryDescription(description, callback) {
        return __awaiter(this, void 0, void 0, function* () {
            this._temporaryDescription = description;
            try {
                yield this.refresh();
                yield callback();
            }
            finally {
                this._temporaryDescription = undefined;
                yield this.refresh();
            }
        });
    }
}
exports.AzureTreeItem = AzureTreeItem;
