// vscode-azureextensionui/out/src/treeDataProvider/GenericTreeItem.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
const AzureTreeItem_1 = require("./AzureTreeItem");
class GenericTreeItem extends AzureTreeItem_1.AzureTreeItem {
    constructor(parent, options) {
        super(parent);
        this.label = options.label;
        this.contextValue = options.contextValue;
        this.id = options.id;
        this.commandId = options.commandId;
        this.iconPath = options.iconPath;
        this.description = options.description;
    }
    isAncestorOfImpl() {
        // never display generic items in tree picker
        return false;
    }
}
exports.GenericTreeItem = GenericTreeItem;
