// microsoft.aspnetcore.razor.vscode/dist/Html/HtmlProjectedDocument.js
"use strict";
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
Object.defineProperty(exports, "__esModule", { value: true });
const UriPaths_1 = require("../UriPaths");
class HtmlProjectedDocument {
    constructor(uri) {
        this.uri = uri;
        this.content = '';
        this.hostDocumentVersion = null;
        this.projectedDocumentVersion = 0;
        this.path = UriPaths_1.getUriPath(uri);
    }
    get hostDocumentSyncVersion() {
        return this.hostDocumentVersion;
    }
    get projectedDocumentSyncVersion() {
        return this.projectedDocumentVersion;
    }
    getContent() {
        return this.content;
    }
    setContent(content, hostDocumentVersion) {
        this.projectedDocumentVersion++;
        this.hostDocumentVersion = hostDocumentVersion;
        this.content = content;
    }
}
exports.HtmlProjectedDocument = HtmlProjectedDocument;
